Change generator `--csv` to `--type=csv`

sambostock opened the pull request:
## TL;DR

```diff
-$ rails generate maintenance_tasks:task NAME --csv
+$ rails generate maintenance_tasks:task NAME --type=csv # or --type csv
```

## Details

Working on #326 and wanting to include a generator for tasks with a custom enumerator builder, @adrianna-chang-shopify and I decided it might be best to change the `--csv` generator option to `--type=csv`, rather than have a bunch of mutually exclusive options (e.g. `--csv` & `--custom`).

IF we decide to provide additional built-in enumerators in the future, this would also provide a natural extension point.

Included in this PR are various refactorings to the generator code.

This is technically a breaking change to the developer API, but not to the actual task/application code. As such, it doesn't feel like it warrants a major version bump.

### ⚠️ Before Merging
- [x] Get feedback on task type names (`--type=generic` & `--type=csv`)
- [x] Update README and any other documentation mentioning `--csv`

adrianna-chang-shopify commented on lib/generators/maintenance_tasks/task_generator.rb:
```diff
@@ -9,23 +9,29 @@ class TaskGenerator < Rails::Generators::NamedBase
 
-    class_option :csv, type: :boolean, default: false,
-      desc: 'Generate a CSV Task.'
+    TASK_TYPES = ['collection', 'csv'].freeze
```
I don't think `.freeze` is necessary 😄

sambostock replied:
I'm just used to freezing all mutable constants (so they're actually "constant"). Do you see a use case for modifying it?

adrianna-chang-shopify replied:
The argument I've heard from Rafael & others is that using `.freeze` is defensive and that we should assume developers are well-intended with the code they write instead of protecting against something that they likely won't do. The fact that it only warns instead of raising if you redefine the constant means that we're not really protected from mutation anyways. We haven't frozen any of our other constants in the gem for this reason 😄

adrianna-chang-shopify commented on lib/generators/maintenance_tasks/task_generator.rb:
```diff
@@ -9,23 +9,29 @@ class TaskGenerator < Rails::Generators::NamedBase
-    class_option :csv, type: :boolean, default: false,
-      desc: 'Generate a CSV Task.'
+    TASK_TYPES = ['collection', 'csv'].freeze
+    private_constant :TASK_TYPES
```
Any reason this needs to be private in addition to the class being private?

sambostock replied:
Habit, since it's really an implementation detail. I can remove it if it's inappropriate.

adrianna-chang-shopify reviewed (commented):
A couple small comments, looks great though!

I am having second thoughts about using `collection` as the default type. (Since `--type=csv` means a collection of type CSV, it's a bit odd to have `--type=collection`). I wonder if it should just be `default` or `generic` instead...

sambostock commented:
@adrianna-chang-shopify I've
- removed `private_constant :TASK_TYPES`
- changed `--type=collection` to `--type=generic`
- changed `--csv` to `--type=csv` in `README.md`

adrianna-chang-shopify reviewed (approved):
I would drop `.freeze` on `TASK_TYPES` still to avoid defensive programming, but other than that this looks great! ❤️

adrianna-chang-shopify commented:
@sambostock @etiennebarrie how would you guys feel about starting off a feature branch for custom enumeration and shipping these incremental PRs into it so that we can release it in completion? `main` has got some patches on it currently that I'll release as a patch version soon, but I've still got a couple fixes to finish. I think it's probably best to not have bits and pieces of the custom enum stuff going out until it's ready in entirety, right?

sambostock commented:
The following PRs exist right now around supporting custom enumeration:

- Ensure cursor is converted into integer as needed (#340)
- Change cursor from bigint to string (#339)
- Change generator `--csv` to `--type=csv`(#343, this PR)
- Introduce Task#enumerator_builder (#326)

#340 would have to be released in a prior release to the one that includes #339, since it's concerned with forwards compatibility while the app is running, so we wouldn't want to include it in a feature branch. Though we'd have to validate #339 will work before merging #340.

#343 I think could technically ship on its own, though I can understand the value of shipping it in the same release as custom enumeration.

Usually I'm opposed to feature branches and prefer to continuously integrate small changes. For instance, we could always ship part of the custom enumerator change as an internal refactor that doesn't break anything.

That said, I've done most of my development for apps, not libraries/gems, so maybe this is different. If we think there's value in a feature branch here, I'm glad to do it. 🤷

adrianna-chang-shopify commented:
I think we're okay to just ship incrementally in that case 👍  I was recalling a previous experience working on the Packwerk gem where a fix had to be shipped ASAP and there was a feature that was half-functional on the main branch and caused all sorts of headaches 😅  But you're right that the most of these changes are internal refactors that should be OK to ship without fear of having incomplete functionality on main.
This one (#343) is totally fine to go out on its own IMO. I was thinking maybe we'd release it as a minor version (since it is a change in the developer API) after releasing a patch (v1.1.1) but I suppose it's okay just to ship all our bug fixes to 1.2 instead.
#340 can go in with that as well, since it's just an internal refactor.

#339, #326 and a PR to add `custom` to our Task Generator can all go out together as the final piece of custom enumeration, once we're sure everything will work and our bases are covered in terms of backwards compatibility + making sure nothing breaks when changing the column type.

So TL;DR ignore my feature branch request, let me see if I can get the patch version out by the end of the day - if not, I think we can just start merging some of these and go straight to 1.2.

adrianna-chang-shopify commented:
@etiennebarrie you'd mentioned preferring to keep `--csv` and support only `--csv` and `--custom` as mutually exclusive options moving forward. I believe that this PR gives us more flexibility, but agree that it might be premature to assume we'll have other Task types beyond CSV and the new custom enumeration stuff. 

Can you 👍  / 👎  this so we either ship or close?

etiennebarrie commented:
Yeah I'm not really in favor of this PR because:
* I'd like to keep `--csv` around
* `--type=generic` can just be the default, no need to add a name for that
* whenever we introduce `--enumerator` (`custom` is a bit unclear I find), we can just deal with the fact that they're exclusive
* I don't think we're going to need more than that

sambostock commented:
Fair enough
